x/protocolpool/keeper/keeper.go

```diff
@@ -90,6 +101,141 @@ func (k Keeper) GetCommunityPool(ctx context.Context) (sdk.Coins, error) {
+	if err != nil {
+		return sdk.Coin{}, fmt.Errorf("error while withdrawing recipient funds for recipient: %s", recipient.String())
+	}
+
+	return withdrawnAmount, nil
+}
+
+func (k Keeper) withdrawRecipientFunds(ctx context.Context, recipient sdk.AccAddress) (sdk.Coin, error) {
+	// get allocated continuous fund
+	fundsAllocated, err := k.RecipientFundDistribution.Get(ctx, recipient)
+	if err != nil {
+		if errors.Is(err, collections.ErrNotFound) {
+			return sdk.Coin{}, types.ErrNoRecipientFund
+		}
+		return sdk.Coin{}, err
+	}
+
+	// Distribute funds to the recipient from pool module account
+	withdrawnAmount := sdk.NewCoin(sdk.DefaultBondDenom, fundsAllocated)
+	err = k.DistributeFromCommunityPool(ctx, sdk.NewCoins(withdrawnAmount), recipient)
+	if err != nil {
+		return sdk.Coin{}, fmt.Errorf("error while distributing funds to the recipient: %s", recipient.String())
+	}
+
+	// reset fund distribution
+	err = k.RecipientFundDistribution.Set(ctx, recipient, math.ZeroInt())
+	if err != nil {
+		return sdk.Coin{}, err
+	}
+	return withdrawnAmount, nil
+}
+
+func (k Keeper) SetToDistribute(ctx context.Context, amount sdk.Coins, addr string) error {
+	authAddr, err := k.authKeeper.AddressCodec().StringToBytes(addr)
+	if err != nil {
+		return err
+	}
+	hasPermission, err := k.HasPermission(ctx, authAddr)
+	if err != nil {
+		return err
+	}
+	if !hasPermission {
+		return sdkerrors.ErrUnauthorized
+	}
+
+	err = k.ToDistribute.Set(ctx, amount.AmountOf(sdk.DefaultBondDenom))
+	if err != nil {
+		return fmt.Errorf("error while setting ToDistribute: %v", err)
+	}
+	return nil
+}
+
+func (k Keeper) HasPermission(ctx context.Context, addr sdk.AccAddress) (bool, error) {
+	authority := k.GetAuthority()
+	authAcc, err := k.authKeeper.AddressCodec().StringToBytes(authority)
+	if err != nil {
+		return false, err
+	}
+	return addr.Equals(sdk.AccAddress(authAcc)), nil
+}
+
+func (k Keeper) iterateAndUpdateFundsDistribution(ctx context.Context, toDistributeAmount math.Int) error {
+	totalPercentageToBeDistributed := math.ZeroInt()
+	err := k.RecipientFundPercentage.Walk(ctx, nil, func(key sdk.AccAddress, value math.Int) (stop bool, err error) {
+		totalPercentageToBeDistributed = totalPercentageToBeDistributed.Add(value)
+		return false, nil
+	})
+	if err != nil {
+		return err
+	}
+	if totalPercentageToBeDistributed.GT(math.NewInt(100)) {
+		return fmt.Errorf("total funds percentage cannot exceed 100")
+	}
+
+	denom := sdk.DefaultBondDenom
+	toDistributeDec := sdk.NewDecCoins(sdk.NewDecCoin(denom, toDistributeAmount))
+
+	// Calculate the funds to be distributed based on the total percentage to be distributed
+	totalAmountToBeDistributed := toDistributeDec.MulDec(math.LegacyNewDecFromIntWithPrec(totalPercentageToBeDistributed, 2))
+	totalDistrAmount := totalAmountToBeDistributed.AmountOf(denom)
+
+	err = k.RecipientFundPercentage.Walk(ctx, nil, func(key sdk.AccAddress, value math.Int) (stop bool, err error) {
```

Comment:
should we store the value in a short lived map in this function to avoid two iterations over the same value? Not sure the performance overhead but seems like double the compute, potentially increasing gas costs as well